libs/libcommon/src/libcommon/orchestrator.py

```diff
@@ -0,0 +1,757 @@
+            + list(self.cache_status.cache_is_outdated_by_parent.values())
+            + list(self.cache_status.cache_is_job_runner_obsolete.values())
+            + list(self.cache_status.cache_has_different_git_revision.values())
+        )
+        for artifact_state in artifact_states:
+            valid_pending_jobs_df = artifact_state.job_state.valid_pending_jobs_df
+            if valid_pending_jobs_df.empty:
+                job_infos_to_create.append(
+                    {
+                        "job_id": "not used",
+                        "type": artifact_state.processing_step.job_type,
+                        "params": {
+                            "dataset": self.dataset,
+                            "revision": self.revision,
+                            "config": artifact_state.config,
+                            "split": artifact_state.split,
+                        },
+                        "priority": self.priority,
+                    }
+                )
+            else:
+                pending_jobs_to_delete_df.drop(valid_pending_jobs_df.index, inplace=True)
+        # Better keep this order: delete, then create
+        # Note that all the pending jobs for other revisions will be deleted
+        if not pending_jobs_to_delete_df.empty:
+            self.add_task(DeleteJobsTask(jobs_df=pending_jobs_to_delete_df))
+        if job_infos_to_create:
+            self.add_task(CreateJobsTask(job_infos=job_infos_to_create))
+
+
+@dataclass
+class DatasetOrchestrator:
+    dataset: str
+    processing_graph: ProcessingGraph
+
+    def set_revision(self, revision: str, priority: Priority, error_codes_to_retry: List[str]) -> None:
+        """
+        Set the current revision of the dataset.
+
+        If the revision is already set to the same value, this is a no-op. Else: one job is created for every first
+          step.
+
+        Args:
+            revision (str): The new revision of the dataset.
+            priority (Priority): The priority of the jobs to create.
+            error_codes_to_retry (List[str]): The error codes for which the jobs should be retried.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: If the first processing steps are not dataset steps, or if the processing graph has no first
+              step.
+        """
+        first_processing_steps = self.processing_graph.get_first_processing_steps()
+        if len(first_processing_steps) < 1:
+            raise ValueError("Processing graph has no first step")
+        if any(first_processing_step.input_type != "dataset" for first_processing_step in first_processing_steps):
+            raise ValueError("One of the first processing steps is not a dataset step")
+        with StepProfiler(
+            method="DatasetOrchestrator.set_revision",
+            step="all",
+            context=f"dataset={self.dataset}",
+        ):
+            logging.info(f"Analyzing {self.dataset}")
```

Comment:
After this line it looks similar logic for the backfill method in line 720, maybe you could call it with different parameters if needed?